### PR TITLE
Fix various ISIS SANS clipboard bugs

### LINKED ISSUE #33286: ISIS SANS Clipboard paste is reversed
**Describe the bug**
When copy and pasting a range of rows, the row order is flipped.

Found in #33195 

**To Reproduce**
- Shift select from top to bottom (or bottom to top)
- Paste into the table
- Notice how the order can be reversed 

**Expected behavior**
- Both should produce the same order matching the current order

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
- Fixes all the edge cases found in #33195 
- Port (copy and paste) the existing implementation into the table model and surround it in tests

See commits for more details

**To test:**
- Download the ISIS Sample Data
- Locate the `loqdemo` folder in the sample data
- Open the ISIS SANS interface
- Click batch file, load the CSV
- Click the row with name `first_row` and shift click select down to `second_row`, click copy
- Click off the rows (so none is selected), click paste and check the order copied correctly
- Do the same steps again, but this time selecting `second_row` first, then `first_row`
- Check the pasted order is goes first->second again
- Try breaking copy / cut paste various different ways 

Fixes #33286 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Looks like another edge case is needed when pasting multiple rows to an empty row when only trailing empty rows exist below?
image

**Comment by DavidFair** [comment]
On discussion me and Richard agreed to push this to 6.4, as it borderlines a new feature
The current feature (whilst clunky) does not cause crashes / problems so we agreed to play it safe

**Review by RichardWaiteSTFC — DISMISSED** [review]
This looks good, behaves like other spreadsheet software which I find quite intuitive. Test coverage looks good for many edge cases.

File: docs/source/release/v6.3.0/33286_sans.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Bugfixes
```
Comment (martyngigg) [inline docs/source/release/v6.3.0/33286_sans.rst:1]: If we're targeting this at 6.4 can we move the release notes to the 6.4 section?

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Comment by RichardWaiteSTFC** [comment]
@DavidFair - sorry forgot about this. Do I need to test it again, or did you just change the release notes in your force push?

**Comment by DavidFair** [comment]
It was a rebase from release-next to main so I could shift the release notes into 6.4
This should be identical code-wise to before

**Comment by RichardWaiteSTFC** [comment]
> @DavidFair - just thought I'd try this for peace of mind. Looks like it's functioning correctly but is very slow, it hangs for ~5s when pasting. Probably this is my computer and there's no reason your code is slowing it down, I was wondering whether you could check?

Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.

**Comment by DavidFair** [comment]
> > @DavidFair - just thought I'd try this for peace of mind. Looks like it's functioning correctly but is very slow, it hangs for ~5s when pasting. Probably this is my computer and there's no reason your code is slowing it down, I was wondering whether you could check?
> 
> Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.

This is because the table opens the associated .nxs file to get the instrument name. Unfortunately it does the `n` times for `n` rows because of a long-standing bug (I can't find the issue # at the moment)

**Review by RichardWaiteSTFC — APPROVED** [review]
> > Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.
> 
> This is because the table opens the associated .nxs file to get the instrument name. Unfortunately it does the `n` times for `n` rows because of a long-standing bug (I can't find the issue # at the moment)

OK I'll approve!

**Review by martyngigg — APPROVED** [review]
